SMQ-2739 - Prevent removing of all users from Built in admin role

JeffMboya opened the pull request:
<!-- Copyright (c) Abstract Machines
SPDX-License-Identifier: Apache-2.0 -->

# What type of PR is this?

This is a feature
## What does this do?

This PR introduces a safeguard that prevents all members from being removed from the built-in `admin` role on any domain. When an attempt is made to remove the last user from the `admin` role, the system now rejects the request with:

```bash
{
    "error": "",
    "message": "failed to remove entity"
}
```

This ensures that at least one user always retains admin privileges, preventing domains from becoming orphaned. 

Example behavior:
- If a domain has 2 admin users, removing one is allowed.
- Attempting to remove both in the same request is rejected.
- If only one admin remains, any attempt to remove them is blocked.

## Which issue(s) does this PR fix/relate to?

- Resolves #2739
- Related # https://github.com/absmach/supermq-docs/pull/223

## Have you included tests for your changes?

Yes. Manual testing was performed by removing members via the `/domains/{domainID}/roles/{roleID}/members/delete` endpoint and verifying behavior with various admin user counts. Unit tests will follow up in a separate test refactor PR.

## Did you document any new/modified feature?

Yes. Documented at https://github.com/absmach/supermq-docs/pull/223

### Notes

JeffMboya committed:
feat(roles): prevent removal of last member from admin role

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

feat(roles): prevent RoleRemoveAllMembers from acting on admin role

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

feat(roles): update implementation

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

feat(roles): fix 404 on RoleRemoveAllMembers

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Fix CI

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

replace hardcoded admin checks with builtInRoles map lookup

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

arvindh123 commented on pkg/roles/provisionmanage.go:
```diff
@@ -559,6 +559,16 @@ func (r ProvisionManageService) RoleRemoveMembers(ctx context.Context, session a
+		if page.Total <= uint64(len(members)) {
+			return errors.Wrap(svcerr.ErrRemoveEntity, fmt.Errorf("built-in role '%s' must retain at least one member", ro.Name))
+		}
+	}
```
This should be applicable only for `domains`
We can move this logic to domain  service `func (svc *service) RoleRemoveAllMembers(ctx context.Context, session authn.Session, entityID, roleID string) error`

JeffMboya committed:
Refactor: Move role guard to domain

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

arvindh123 commented on domains/service.go:
```diff
@@ -351,15 +352,75 @@ func (svc *service) RemoveEntityMembers(ctx context.Context, session authn.Sessi
+				return errors.Wrap(svcerr.ErrRemoveEntity, fmt.Errorf("failed to delete invitation for member '%s' in domain '%s' from role '%s': %w", memberID, entityID, ro.Name, err))
+			}
+		}
+	}
```
This logic is taken care by `RoleRemoveAllMembers`
Please remove this logic

arvindh123 commented on domains/service.go:
```diff
@@ -351,15 +352,75 @@ func (svc *service) RemoveEntityMembers(ctx context.Context, session authn.Sessi
+	if _, ok := svc.ProvisionManageService.BuiltInRoles[roles.BuiltInRoleName(ro.Name)]; ok {
+		isBuiltInRole = true
+	}
+
```
We can simply have like below
```suggestion
	if _, ok := svc.ProvisionManageService.BuiltInRoles[roles.BuiltInRoleName(ro.Name)]; ok {
		return errors.Wrap(svcerr.ErrRemoveEntity, fmt.Errorf("built-in role '%s' must retain at least one member and cannot have all members removed", ro.Name))
	}

```

arvindh123 commented on domains/service.go:
```diff
@@ -351,15 +352,75 @@ func (svc *service) RemoveEntityMembers(ctx context.Context, session authn.Sessi
+		if membersPage.Total <= uint64(len(members)) {
+			return errors.Wrap(svcerr.ErrRemoveEntity, fmt.Errorf("built-in role '%s' must retain at least one member. Attempting to remove %d member(s) from %d would leave too few.", ro.Name, len(members), membersPage.Total))
 		}
 	}
```
```suggestion
	if _, ok := svc.ProvisionManageService.BuiltInRoles[roles.BuiltInRoleName(ro.Name)]; ok {

		membersPage, listErr := svc.repo.RoleListMembers(ctx, ro.ID, 0, 0)
		if listErr != nil {
			return errors.Wrap(svcerr.ErrViewEntity, fmt.Errorf("failed to list members for built-in role '%s' to check count: %w", ro.Name, listErr))
		}

		if membersPage.Total <= uint64(len(members)) {
			return errors.Wrap(svcerr.ErrRemoveEntity, fmt.Errorf("built-in role '%s' must retain at least one member. Attempting to remove %d member(s) from %d would leave too few.", ro.Name, len(members), membersPage.Total))
		}
	}
```

arvindh123 commented on domains/service.go:
```diff
@@ -351,15 +352,75 @@ func (svc *service) RemoveEntityMembers(ctx context.Context, session authn.Sessi
 	}
+
+	for _, memberID := range members {
+		if err := svc.repo.DeleteInvitation(ctx, memberID, entityID); err != nil && err != repoerr.ErrNotFound {
```
We can have repo call where we can delete All the Invitation at once something like 

```go 
svc.repo.DeleteUsersInvitation(ctx,entityID, memberIDs...)
```

This requires new function invitations repo like ```DeleteUsersInvitation(ctx,entityID, memberIDs...)```

arvindh123 replied:
Create ticket for this https://github.com/absmach/supermq/issues/2924
We can solve this later in different PR

JeffMboya committed:
Address comments

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Address comments

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Address comments

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Address comments

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Address comments

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Refactor error

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Fix CI

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Update domains/service.go

Co-authored-by: Arvindh <30824765+arvindh123@users.noreply.github.com>

Fix CI

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Refactor

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Update error

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Update error

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

UFix CI

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Refactor

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Refactor

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Refactor

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Fix CI

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Refactor

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

arvindh123 commented on domains/service.go:
```diff
@@ -351,15 +349,50 @@ func (svc *service) RemoveEntityMembers(ctx context.Context, session authn.Sessi
+	ro, err := svc.repo.RetrieveEntityRole(ctx, entityID, roleID)
+	if err != nil {
+		return errors.Wrap(svcerr.ErrViewEntity, fmt.Errorf("failed to retrieve role '%s' for entity '%s': %w", roleID, entityID, err))
+	}
```
```suggestion
		return errors.Wrap(svcerr.ErrViewEntity, err)
```

arvindh123 commented on domains/service.go:
```diff
@@ -351,15 +349,50 @@ func (svc *service) RemoveEntityMembers(ctx context.Context, session authn.Sessi
+		membersPage, listErr := svc.repo.RoleListMembers(ctx, ro.ID, 0, 0)
+		if listErr != nil {
+			return errors.Wrap(svcerr.ErrViewEntity, fmt.Errorf("failed to list members for built-in role '%s' to check count: %w", ro.Name, listErr))
+		}
```
```suggestion
			return errors.Wrap(svcerr.ErrViewEntity, err)
```

arvindh123 replied:
```suggestion
		membersPage, err := svc.repo.RoleListMembers(ctx, ro.ID, 0, 0)
		if err != nil {
```

arvindh123 commented on domains/service.go:
```diff
@@ -351,15 +349,50 @@ func (svc *service) RemoveEntityMembers(ctx context.Context, session authn.Sessi
+		}
+		if membersPage.Total <= uint64(len(members)) {
+			return errors.Wrap(svcerr.ErrRemoveEntity, fmt.Errorf("built-in role '%s' must retain at least one member. Attempting to remove %d member(s) from %d would leave too few.", ro.Name, len(members), membersPage.Total))
+		}
```
```suggestion
errAtLeastOneMemberRetain = errors.New("at least one member should be in inbuild role")
			return errors.Wrap(svcerr.ErrRemoveEntity, errAtLeastOneMemberRetain)
```

No need for details , like how many members request and role name

arvindh123 commented on domains/service.go:
```diff
@@ -351,15 +349,50 @@ func (svc *service) RemoveEntityMembers(ctx context.Context, session authn.Sessi
+	ro, err := svc.repo.RetrieveEntityRole(ctx, entityID, roleID)
+	if err != nil {
+		return errors.Wrap(svcerr.ErrViewEntity, fmt.Errorf("failed to retrieve role '%s' for entity '%s': %w", roleID, entityID, err))
+	}
```
```suggestion
		return errors.Wrap(svcerr.ErrViewEntity, err)
```

arvindh123 commented on domains/service.go:
```diff
@@ -351,15 +349,50 @@ func (svc *service) RemoveEntityMembers(ctx context.Context, session authn.Sessi
+	if _, ok := svc.ProvisionManageService.BuiltInRoles[roles.BuiltInRoleName(ro.Name)]; ok {
+		membersPage, listErr := svc.repo.RoleListMembers(ctx, ro.ID, 0, 0)
+		if listErr != nil {
+			return errors.Wrap(svcerr.ErrViewEntity, fmt.Errorf("failed to list members for built-in role '%s' to check count: %w", ro.Name, listErr))
```
```suggestion
		membersPage, err := svc.repo.RoleListMembers(ctx, ro.ID, 0, 0)
		if err != nil {
```

arvindh123 commented on domains/service.go:
```diff
@@ -351,15 +349,50 @@ func (svc *service) RemoveEntityMembers(ctx context.Context, session authn.Sessi
+		membersPage, listErr := svc.repo.RoleListMembers(ctx, ro.ID, 0, 0)
+		if listErr != nil {
+			return errors.Wrap(svcerr.ErrViewEntity, fmt.Errorf("failed to list members for built-in role '%s' to check count: %w", ro.Name, listErr))
+		}
```
```suggestion
			return errors.Wrap(svcerr.ErrViewEntity, err)
```

arvindh123 commented on domains/service.go:
```diff
@@ -351,15 +349,50 @@ func (svc *service) RemoveEntityMembers(ctx context.Context, session authn.Sessi
+			return errors.Wrap(svcerr.ErrViewEntity, fmt.Errorf("failed to list members for built-in role '%s' to check count: %w", ro.Name, listErr))
+		}
+		if membersPage.Total > 0 {
+			return errors.Wrap(svcerr.ErrRemoveEntity, fmt.Errorf("built-in role '%s' must retain at least one member and cannot have all members removed", ro.Name))
```
No need  of details 
Just follow the approach mentioned here https://github.com/absmach/supermq/pull/2835#discussion_r2129125671

arvindh123 commented on domains/service.go:
```diff
@@ -157,7 +156,6 @@ func (svc service) DisableDomain(ctx context.Context, session authn.Session, id
 	return dom, nil
 }
 
-// Only SuperAdmin can freeze the domain.
```
Why this comment is removed ?

arvindh123 commented on domains/service.go:
```diff
@@ -63,7 +63,6 @@ func (svc service) CreateDomain(ctx context.Context, session authn.Session, d Do
 
 	d.CreatedAt = time.Now().UTC()
 
-	// Domain is created in repo first, because Roles table have foreign key relation with Domain ID
```
Why this comment is removed ?

arvindh123 commented on domains/service.go:
```diff
@@ -341,7 +339,6 @@ func (svc *service) DeleteInvitation(ctx context.Context, session authn.Session,
 	return nil
 }
 
-// Add addition removal of user from invitations.
```
Why this comment is removed ?

arvindh123 commented on domains/service.go:
```diff
@@ -351,15 +352,50 @@ func (svc *service) RemoveEntityMembers(ctx context.Context, session authn.Sessi
+			return errors.Wrap(svcerr.ErrViewEntity, err)
+		}
+		if membersPage.Total <= uint64(len(members)) {
+			return errors.Wrap(svcerr.ErrRemoveEntity, errAtLeastOneMemberRetain)
```
@JeffMboya 

I’ve described the general approach in [this comment](https://github.com/absmach/supermq/pull/2835#discussion_r2129125671), but it shouldn’t be taken as a literal implementation.

Also, as per our established coding conventions, all error constants must be defined at the top of the file. Please make sure to move it there.

arvindh123 commented on domains/service.go:
```diff
@@ -351,15 +352,50 @@ func (svc *service) RemoveEntityMembers(ctx context.Context, session authn.Sessi
+			return errors.Wrap(svcerr.ErrViewEntity, err)
+		}
+		if membersPage.Total > 0 {
+			return errors.Wrap(svcerr.ErrRemoveEntity, errAtLeastOneMemberRetain)
```
@JeffMboya 

I’ve described the general approach in [this comment](https://github.com/absmach/supermq/pull/2835#discussion_r2129125671), but it shouldn’t be taken as a literal implementation.

Also, as per our established coding conventions, all error constants must be defined at the top of the file. Please make sure to move it there.

Please try to reuse the same variable

JeffMboya replied:
@arvindh123 yes, it is already implemented in the last commit

arvindh123 commented on domains/service.go:
```diff
@@ -19,6 +19,7 @@ import (
 var (
 	errCreateDomainPolicy = errors.New("failed to create domain policy")
 	errRollbackRepo       = errors.New("failed to rollback repo")
+	errAtLeastOneMemberRetain = errors.New("at least one member must remain in built-in role")
```
```suggestion
	errCreateDomainPolicy     = errors.New("failed to create domain policy")
	errRollbackRepo           = errors.New("failed to rollback repo")
	errAtLeastOneMemberRetain = errors.New("at least one member must remain in built-in role")
```

Please algin them

arvindh123 commented on pkg/roles/provisionmanage.go:
```diff
@@ -34,7 +34,7 @@ type ProvisionManageService struct {
 	policy       policies.Service
 	actions      []Action
-	builtInRoles map[BuiltInRoleName][]Action
+	BuiltInRoles map[BuiltInRoleName][]Action
```
Instead of exporting the field `builtInRoles` , please add function like `func (pms ProvisionManageService) BuiltInRoleActions(name BuiltInRoleName)([]Action, error)`
and in `domains/service.go`, we can have something like below 
```go
if _, err := svc.ProvisionManageService.BuiltInRoleActions; err == nil {
    membersPage, err := svc.repo.RoleListMembers(ctx, ro.ID, 0, 0)
    if err != nil {
        return errors.Wrap(svcerr.ErrViewEntity, err)
    }
    if membersPage.Total <= uint64(len(members)) {
        return errors.Wrap(svcerr.ErrRemoveEntity, errAtLeastOneMemberRetain)
    }
}
```

arvindh123 commented on domains/service.go:
```diff
@@ -351,15 +352,50 @@ func (svc *service) RemoveEntityMembers(ctx context.Context, session authn.Sessi
+			return errors.Wrap(svcerr.ErrViewEntity, err)
+		}
+		if membersPage.Total <= uint64(len(members)) {
+			return errors.Wrap(svcerr.ErrRemoveEntity, errAtLeastOneMemberRetain)
```
@JeffMboya  

API response code is 400 and In response we have no details for failure 
```json
{
    "error": "",
    "message": "view entity failed"
}
```
So please move the error `errAtLeastOneMemberRetain` to `pkg/errors/service/types.go` 
and use that like `svcerr.ErrAtLeastOneMemberRetain`

Please add this `svcerr.ErrAtLeastOneMemberRetain` to `EncodeError` function at `api/http/common.go`
for switch case `http.StatusBadRequest` 
https://github.com/absmach/supermq/blob/b79fdb15343a43b0db45153bdfa4102053f1027f/api/http/common.go#L238

JeffMboya replied:
Updated:

```bash
{
    "error": "",
    "message": "must retain at least one member"
}
```

arvindh123 commented on domains/service.go:
```diff
@@ -351,15 +351,50 @@ func (svc *service) RemoveEntityMembers(ctx context.Context, session authn.Sessi
+			return errors.Wrap(svcerr.ErrViewEntity, err)
+		}
+		if membersPage.Total <= uint64(len(members)) {
+			return errors.Wrap(svcerr.ErrRetainOneMember, err)
```
@JeffMboya 
Do you think here we need to wrap with `err`?

JeffMboya replied:
> @JeffMboya Do you think here we need to wrap with `err`?

Not really. Let me update

arvindh123 reviewed (changes requested):
@Jeff

arvindh123 commented on domains/service.go:
```diff
@@ -351,15 +351,50 @@ func (svc *service) RemoveEntityMembers(ctx context.Context, session authn.Sessi
+	if len(members) == 0 {
+		return svcerr.ErrMalformedEntity
+	}
+
```
@JeffMboya 
Validation is done in API layer, So we can remove validation here

arvindh123 commented on domains/service.go:
```diff
@@ -351,15 +351,46 @@ func (svc *service) RemoveEntityMembers(ctx context.Context, session authn.Sessi
+		}
+		if membersPage.Total > 0 {
+			return svcerr.ErrRetainOneMember
+		}
```
Here simply we can return `svcerr.ErrRetainOneMember` if the role is BuiltInRole
No need for checking the members total. 
Because this function itself going to remove everyone.

arvindh123 commented on domains/service.go:
```diff
@@ -351,15 +351,40 @@ func (svc *service) RemoveEntityMembers(ctx context.Context, session authn.Sessi
+
+	for _, memberID := range members {
+		if err := svc.repo.DeleteInvitation(ctx, memberID, entityID); err != nil && err != repoerr.ErrNotFound {
+			return svcerr.ErrRetainOneMember
```
Please revert this error, Here we should send the error returned by `svc.repo.DeleteInvitation`

JeffMboya committed:
Refactor

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>
